_posts/2022-03-16-async-sink-base.md

```diff
@@ -0,0 +1,169 @@
+
+{% toc %}
+
+# Adding the base sink as a dependency
+
+In order to use the base sink, you will need to add the following dependency to your project. The example below follows the Maven syntax:
+
+```xml
+<dependency>
+  <groupId>org.apache.flink</groupId>
+  <artifactId>flink-connector-base</artifactId>
+  <version>${flink.version}</version>
+</dependency>
+```
+
+# The Public Interfaces of AsyncSinkBase
+
+## Generic Types
+
+`<InputT>` – type of elements in a DataStream that should be passed to the sink
+
+`<RequestEntryT>` – type of a payload containing the element and additional metadata that is required to submit a single element to the destination
+
+
+## Element Converter Interface
+
+[ElementConverter](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/ElementConverter.java)
+
+```java
+public interface ElementConverter<InputT, RequestEntryT> extends Serializable {
+    RequestEntryT apply(InputT element, SinkWriter.Context context);
+}
+```
+The concrete sink implementation should provide a way to convert from an element in the DataStream to the payload type that contains all the additional metadata required to submit that element to the destination by the sink. Ideally, this would be encapsulated from the end user since it allows concrete sink implementers to adapt to changes in the destination API without breaking end user code.
+
+## Sink Writer Interface
+
+[AsyncSinkWriter](https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/sink/writer/AsyncSinkWriter.java)
+
+```java
+public abstract class AsyncSinkWriter<InputT, RequestEntryT extends Serializable>
+        implements StatefulSink.StatefulSinkWriter<InputT, BufferedRequestState<RequestEntryT>> {
+    // ...
+    protected abstract void submitRequestEntries(
+            List<RequestEntryT> requestEntries, Consumer<List<RequestEntryT>> requestResult);
+    // ...
+}
+```
+
+In this method, sink implementers should use the destination clients to submit `requestEntries` asynchronously to be written.
+
+Should any elements fail to be persisted, they should be requeued back in the buffer for retry using `requestResult.accept(...list of failed entries...)`. However, retrying any element that is known to be faulty and consistently failing, will result in that element being requeued forever, therefore a sensible strategy for determining what should be retried is highly recommended. If no errors were returned, we must indicate this with `requestResult.accept(Collections.emptyList())`.
+
+If at any point, it is determined that a fatal error has occurred and that we should throw a runtime exception from the sink, we can call `getFatalExceptionCons().accept(...);` from anywhere in the concrete sink writer.
+
+```java
+public abstract class AsyncSinkWriter<InputT, RequestEntryT extends Serializable>
```

Comment:
What's the migration story here? By using java serialization you're locking the implementation to a specific class. How do you ensure that states can be migrated from previous versions? How can connectors change the connector-internal state representation?

I hope these issues were raised during the reviews; we do call out java serialization in our code quality guide (repeatedly) for a reason. Any usage of java serialization in api & state layers have proven to be problematic down-the-line, even when they sounded like a good idea at the time.

Reply:
Java serialization is not actually used in the `AsyncSinkWriterStateSerializer`. It was in the original PR, which was rejected and subsequently reimplemented. Looks like `extends Serializable` was then not removed, which it should have been. This is a smell in the interface, since this is `@PublicEvolving` we can consider removing this in Flink 1.16 with community buy-in.